File: DDD DSL/DDD DSL Design War Measurement System/Code/Java/src/fayelab/ddd/measure/vector/scale/MeasureSystemUI.java
package fayelab.ddd.measure.vector.scale;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static java.util.Arrays.asList;

public class MeasureSystemUI
{
    @SuppressWarnings("unused")
    private String sysName;

    private MeasureSystem ms;
    private List<String> sysUnits;

    @SuppressWarnings("unchecked")
    public MeasureSystemUI(String sysName, String unitConversionDesc)
    {
        this.sysName = sysName;
        List<List<?>> unitConversion = parseUnitConversionDesc(unitConversionDesc);
        this.ms = new MeasureSystem((List<Integer>)unitConversion.get(0));
        this.sysUnits = (List<String>)unitConversion.get(1);
    }

    static List<List<?>> parseUnitConversionDesc(String desc)
    {
        Map<Boolean, List<String>> partitioned = partitioningByIndexParity(tokenize(desc));

        List<Integer> stepFactors = toIntegers(partitioned.get(Boolean.FALSE));
        stepFactors.add(0, Integer.MAX_VALUE);
        List<String> units = partitioned.get(Boolean.TRUE);

        return asList(stepFactors, units);
    }

    @SuppressWarnings("unchecked")
    static List<List<?>> parseUnitConversionDesc2(String desc)
    {
        List<String> tokens = tokenize(desc);

        return tokens.stream()
                     .collect(() -> asList(new ArrayList<Integer>(asList(Integer.MAX_VALUE)), new ArrayList<String>()),
                              (acc, token) -> {
                                  if(isEvenIndex(token, tokens))
                                  {
                                      ((List<String>)acc.get(1)).add(token);
                                  }
                                  else
                                  {
                                      ((List<Integer>)acc.get(0)).add(Integer.parseInt(token));
                                  }
                              },
                              (acc1, acc2) -> {
                                  ((List<Integer>)acc1.get(0)).addAll((List<Integer>)acc2.get(0));
                                  ((List<String>)acc1.get(1)).addAll((List<String>)acc2.get(1));
                              });
    }

    List<Integer> parseQuantityDesc(String desc)
    {
        Map<Boolean, List<String>> partitioned = partitioningByIndexParity(tokenize(desc));

        List<Integer> values = toIntegers(partitioned.get(Boolean.TRUE));
        List<String> units = partitioned.get(Boolean.FALSE);

        return sysUnits.stream()
                       .map(sysUnit -> units.contains(sysUnit) ? values.get(units.indexOf(sysUnit)) : 0)
                       .collect(Collectors.toList());
    }

    public boolean equal(String quantityDesc1, String quantityDesc2)
    {
        return ms.equal(parseQuantityDesc(quantityDesc1), parseQuantityDesc(quantityDesc2));
    }

    public String format(List<Integer> quantityVec)
    {
        return IntStream.range(0, quantityVec.size())
                        .filter(i -> quantityVec.get(i) != 0 || i == quantityVec.size() - 1)
                        .mapToObj(i -> String.format("%d %s", quantityVec.get(i), sysUnits.get(i)))
                        .reduce((acc, valueAndUnit) -> String.join(" ", acc, valueAndUnit))
                        .orElse("");
    }

    public String add(String quantityDesc1, String quantityDesc2)
    {
        return format(ms.add(parseQuantityDesc(quantityDesc1), parseQuantityDesc(quantityDesc2)));
    }

    public String baseFormat(List<Integer> quantityVec)
    {
        return String.format("%d %s", ms.base(quantityVec), sysUnits.get(sysUnits.size() - 1));
    }

    public String scale(String c, String quantityDesc)
    {
        return format(ms.scale(Integer.parseInt(c), parseQuantityDesc(quantityDesc)));
    }

    private static List<String> tokenize(String desc)
    {
        return asList(desc.trim().split(" +"));
    }

    private static Map<Boolean, List<String>> partitioningByIndexParity(List<String> tokens)
    {
        return tokens.stream().collect(Collectors.partitioningBy(token -> isEvenIndex(token, tokens)));
    }

    private static boolean isEvenIndex(String token, List<String> tokens)
    {
        return tokens.indexOf(token) % 2 == 0;
    }

    private static List<Integer> toIntegers(List<String> strs)
    {
        return strs.stream().map(Integer::parseInt).collect(Collectors.toList());
    }

    public static void main(String[] args)
    {
        MeasureSystemUI impLen = new MeasureSystemUI("Imperial Length", "Mile 1760 Yard 3 Feet 12 Inch");
        System.out.println("1765 Yard 40 Inch == 1 Mile 6 Yard 4 Inch ? " + impLen.equal("1765 Yard 40 Inch", "1 Mile 6 Yard 4 Inch"));
        System.out.println("13 Inch + 11 Inch = " + impLen.add("13 Inch", "11 Inch"));

        MeasureSystemUI impVol = new MeasureSystemUI("Imperial Volume", "OZ 2 TBSP 3 TSP");
        System.out.println("1 OZ 10 TSP == 2 OZ 1 TBSP 1 TSP ? " + impVol.equal("1 OZ 10 TSP", "2 OZ 1 TBSP 1 TSP"));
        System.out.println("1 OZ + 3 TBSP 3 TSP = " + impVol.add("1 OZ", "3 TBSP 3 TSP"));

        MeasureSystemUI time = new MeasureSystemUI("Time", "Day 24 Hour 60 Minute 60 Second");
        System.out.println("7 Hour 59 Minute 60 Second == 8 Hour ? " + time.equal("7 Hour 59 Minute 60 Second", "8 Hour"));
        System.out.println("1 Day 23 Hour 59 Minute 59 Second + 1 Second = " + time.add("1 Day 23 Hour 59 Minute 59 Second", "1 Second"));

        MeasureSystemUI impLen2 = new MeasureSystemUI("Imperial Length 2", "Mile 4 Miya 440 Yard 3 Feet 12 Inch");
        System.out.println("4 Miya 446 Yard 40 Inch == 1 Mile 1 Miya 7 Yard 4 Inch ? " + impLen2.equal("4 Miya 446 Yard 40 Inch", "1 Mile 1 Miya 7 Yard 4 Inch"));
        System.out.println("3 Miya 13 Inch + 445 Yard 11 Inch = " + impLen2.add("3 Miya 13 Inch", "445 Yard 11 Inch"));

        System.out.println("2 * 2 Yard 3 Feet = " + impLen.scale("2", "2 Yard 3 Feet"));
        System.out.println("2 * 2 Yard 3 Feet + 13 Inch = " + impLen.add(impLen.scale("2", "2 Yard 3 Feet"), "13 Inch"));

        System.out.println("3 * 1 OZ 10 TSP = " + impVol.scale("3", "1 OZ 10 TSP"));
        System.out.println("3 * 1 OZ 10 TSP + 3 TBSP 1 TSP = " + impVol.add(impVol.scale("3", "1 OZ 10 TSP"), "3 TBSP 1 TSP"));
    }
}

File: DDD DSL/DDD DSL Design War Measurement System/Code/Python/vector/scale/measuresystem_ui.py
import re
import string

import measuresystem as ms

PLACEHOLDER = 0x7FFFFFFF

def parse_unitconversion_desc(desc):
    (sysunits, stepfactorstrs) = partitioning_by_index_parity(tokenize(desc))
    stepfactors = ints(stepfactorstrs)
    stepfactors.insert(0, PLACEHOLDER)
    return stepfactors, sysunits

def parse_unitconversion_desc2(desc):
    tokens = tokenize(desc)
    (sysunits, stepfactorstrs, tokens) = reduce(_acc_parse_desc, tokens, ([], [], tokens))
    stepfactors = ints(stepfactorstrs)
    stepfactors.insert(0, PLACEHOLDER)
    return stepfactors, sysunits

def parse_quantity_desc(desc, sysunits):
    (valuestrs, units) = partitioning_by_index_parity(tokenize(desc))
    values = ints(valuestrs)
    return [values[units.index(sysunit)] if sysunit in units else 0 for sysunit in sysunits]

def parse_quantity_desc2(desc, sysunits):
    tokens = tokenize(desc)
    (valuestrs, units, tokens) = reduce(_acc_parse_desc, tokens, ([], [], tokens))
    values = ints(valuestrs)
    return [values[units.index(sysunit)] if sysunit in units else 0 for sysunit in sysunits]

def equal(quantitydesc1, quantitydesc2, sysunits, stepfactors):
    return ms.equal(parse_quantity_desc(quantitydesc1, sysunits),
                    parse_quantity_desc(quantitydesc2, sysunits),
                    stepfactors)

def format(quantityvec, sysunits):
    valueunits = [' '.join([str(value), unit])
                  for (value, unit) in zip(quantityvec, sysunits)
                  if value != 0 or unit == sysunits[-1]]
    return ' '.join(valueunits)

def add(quantitydesc1, quantitydesc2, sysunits, stepfactors):
    return format(ms.add(parse_quantity_desc(quantitydesc1, sysunits),
                         parse_quantity_desc(quantitydesc2, sysunits), stepfactors),
                  sysunits)

def base_format(quantityvec, sysunits, stepfactors):
    basevalue = str(ms.base(quantityvec, ms.basefactors(stepfactors)))
    return ' '.join([basevalue, sysunits[-1]])

def scale(c, quantitydesc, sysunits, stepfactors):
    return format(ms.scale(int(c), parse_quantity_desc(quantitydesc, sysunits), stepfactors), sysunits)

def tokenize(desc):
    return re.split(r' +', string.strip(desc))

def partitioning_by_index_parity(tokens):
    part1 = []
    part2 = []
    for token in tokens:
        part1.append(token) if evenindex(token, tokens) else part2.append(token)
    return part1, part2

def evenindex(token, tokens):
    return tokens.index(token) % 2 == 0

def ints(strs):
    return [int(s) for s in strs]

def _acc_parse_desc((part1, part2, tokens), token):
    part1.append(token) if evenindex(token, tokens) else part2.append(token)
    return part1, part2, tokens

if __name__ == '__main__':
    implen_stepfactors, implen_sysunits = parse_unitconversion_desc('Mile 1760 Yard 3 Feet 12 Inch')
    r1 = equal("1765 Yard 40 Inch", "1 Mile 6 Yard 4 Inch", implen_sysunits, implen_stepfactors)
    print '1765 Yard 40 Inch == 1 Mile 6 Yard 4 Inch ? %s' % (r1)
    r2 = add('13 Inch', '11 Inch', implen_sysunits, implen_stepfactors)
    print '13 Inch + 11 Inch = %s' % (r2)

    impvol_stepfactors, impvol_sysunits = parse_unitconversion_desc('OZ 2 TBSP 3 TSP')
    r3 = equal('1 OZ 10 TSP', '2 OZ 1 TBSP 1 TSP', impvol_sysunits, impvol_stepfactors)
    print '1 OZ 10 TSP == 2 OZ 1 TBSP 1 TSP ? %s' % (r3)
    r4 = add('1 OZ', '3 TBSP 3 TSP', impvol_sysunits, impvol_stepfactors)
    print '1 OZ + 3 TBSP 3 TSP = %s' % (r4)

    time_stepfactors, time_sysunits = parse_unitconversion_desc('Day 24 Hour 60 Minute 60 Second')
    r5 = equal('7 Hour 59 Minute 60 Second', '8 Hour', time_sysunits, time_stepfactors)
    print '7 Hour 59 Minute 60 Second == 8 Hour ? %s' % (r5)
    r6 = add('1 Day 23 Hour 59 Minute 59 Second', '1 Second', time_sysunits, time_stepfactors)
    print '1 Day 23 Hour 59 Minute 59 Second + 1 Second = %s' % (r6)

    implen2_stepfactors, implen2_sysunits = parse_unitconversion_desc('Mile 4 Miya 440 Yard 3 Feet 12 Inch')
    r7 = equal('4 Miya 446 Yard 40 Inch', '1 Mile 1 Miya 7 Yard 4 Inch', implen2_sysunits, implen2_stepfactors)
    print '4 Miya 446 Yard 40 Inch == 1 Mile 1 Miya 7 Yard 4 Inch ? %s' % (r7)
    r8 = add('3 Miya 13 Inch', '445 Yard 11 Inch', implen2_sysunits, implen2_stepfactors)
    print '3 Miya 13 Inch + 445 Yard 11 Inch = %s' % (r8)

    r9 = scale('2', '2 Yard 3 Feet', implen_sysunits, implen_stepfactors)
    print '2 * 2 Yard 3 Feet = %s' % (r9)
    r10 = add(scale('2', '2 Yard 3 Feet', implen_sysunits, implen_stepfactors), '13 Inch', implen_sysunits, implen_stepfactors)
    print '2 * 2 Yard 3 Feet + 13 Inch = %s' % (r10)

    r11 = scale('3', '1 OZ 10 TSP', impvol_sysunits, impvol_stepfactors)
    print '3 * 1 OZ 10 TSP = %s' % (r11)
    r12 = add(scale('3', '1 OZ 10 TSP', impvol_sysunits, impvol_stepfactors), '3 TBSP 1 TSP', impvol_sysunits, impvol_stepfactors)
    print '3 * 1 OZ 10 TSP + 3 TBSP 1 TSP = %s' % (r12)